rust/otap-dataflow/crates/engine/src/retry_processor.rs

```diff
@@ -0,0 +1,788 @@
+
+        for (&id, pending) in &self.pending_messages {
+            if pending.next_retry_time <= now {
+                ready_messages.push((id, pending.data.clone()));
+            }
+        }
+
+        for (id, data) in ready_messages {
+            log::debug!("Retrying message with ID: {id}");
+            effect_handler.send_message(data).await?;
+        }
+
+        Ok(())
+    }
+
+    fn cleanup_expired_messages(&mut self) {
+        let now = Instant::now();
+        if now.duration_since(self.last_cleanup_time)
+            < Duration::from_secs(self.config.cleanup_interval_secs)
+        {
+            return;
+        }
+
+        // Clean up messages that have exceeded max retries and are old
+        let max_age = Duration::from_secs(MAX_FAILED_MESSAGE_AGE_SECS);
+        let expired_ids: Vec<u64> = self
+            .pending_messages
+            .iter()
+            .filter_map(|(&id, pending)| {
+                let age = now.duration_since(pending.next_retry_time);
+                if pending.retry_count > self.config.max_retries && age > max_age {
+                    Some(id)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        for id in expired_ids {
+            if self.pending_messages.remove(&id).is_some() {
+                log::warn!("Removed expired message with ID: {id}");
+            }
+        }
+
+        self.last_cleanup_time = now;
+    }
+
+    fn add_message_for_retry(&mut self, data: PData) -> Result<u64, String> {
+        if self.pending_messages.len() >= self.config.max_pending_messages {
+            return Err(format!(
+                "Retry queue is full (capacity: {}), cannot add message",
+                self.config.max_pending_messages
+            ));
+        }
+
+        let id = self.next_message_id;
+        self.next_message_id += 1;
+
+        let pending = PendingMessage {
+            data,
+            retry_count: 0,
+            next_retry_time: Instant::now(),
+            last_error: String::new(),
+        };
+
+        let _previous = self.pending_messages.insert(id, pending);
+        log::debug!("Added message {id} to retry queue");
+        Ok(id)
+    }
+}
+
+#[async_trait(?Send)]
+impl<PData: Clone + Send + 'static> Processor<PData> for RetryProcessor<PData> {
+    async fn process(
+        &mut self,
+        msg: Message<PData>,
+        effect_handler: &mut EffectHandler<PData>,
+    ) -> Result<(), Error<PData>> {
+        match msg {
+            Message::PData(data) => {
+                match self.add_message_for_retry(data.clone()) {
+                    Ok(_id) => {
+                        effect_handler.send_message(data).await?;
+                    }
+                    Err(error_msg) => {
+                        log::warn!("{error_msg}");
+                        effect_handler.send_message(data).await?;
+                    }
+                }
+                Ok(())
```

Comment:
If I understand this correctly, `add_message_for_retry` returns an error if the retry queue is full right? That means if the retry queue is full, we just send the message without a retry?

I'm wondering if this is actually the behaviour we want, or if instead it makes sense here to send a `Nack` control message which iirc could be used as a form of back pressure?